src/btf_parser.cpp

```diff
@@ -1081,3 +1085,393 @@ std::map<std::string, size_t> parse_btf_map_sections(const btf_type_data& btf_da
     }
     return map_offset_to_descriptor_index;
 }
+
+/**
+ * @brief Write a type into a vector and grow it.
+ *
+ * @tparam T Type to write
+ * @param[in] btf The .BTF section
+ * @param[in] value The value to write
+ */
+template <typename T>
+void _write_btf(std::vector<std::byte>& btf, const T& value) {
+    size_t length = 0;
+    size_t offset = btf.size();
+    if constexpr (std::is_same<T, std::string>::value) {
+        length = value.length();
+        btf.resize(offset + length + 1);
+        memcpy(btf.data() + offset, value.c_str(), length + 1);
+    } else if constexpr (std::is_same<T, std::vector<std::byte>>::value) {
+        length = value.size();
+        btf.resize(offset + length);
+        memcpy(btf.data() + offset, value.data(), length);
+    } else {
+        length = sizeof(T);
+        btf.resize(offset + length);
+        memcpy(btf.data() + offset, &value, length);
+    }
+}
```

Comment:
What I meant is, here you implement function overloading manually. Instead something like the following is more obvious IMO:
```cpp
/**
 * @brief Write a type into a vector and grow it.
 *
 * @param[in,out] btf The .BTF section
 * @param[in] value The value to write
 */
void _write_btf(std::vector<std::byte>& btf, const std::string& value) {
    size_t length = value.length();
    size_t offset = btf.size();
    btf.resize(offset + length + 1);
    memcpy(btf.data() + offset, value.c_str(), length + 1);
}

void _write_btf(std::vector<std::byte>& btf, const std::vector<std::byte>& value) {
    size_t length = value.size();
    size_t offset = btf.size();
    btf.resize(offset + length);
    memcpy(btf.data() + offset, value.data(), length);
}

void _write_btf(std::vector<std::byte>& btf, const auto& value) {
    size_t length = sizeof(value);
    size_t offset = btf.size();
    btf.resize(offset + length);
    memcpy(btf.data() + offset, &value, length);
}
```